mautic/mautic
# Campaign summary statistics (performance boost and date range specificity)

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | 
| New feature? | Yes
| Automated tests included? |
| Related user documentation PR URL | 
| Related developer documentation PR URL | 
| Issues addressed (#s or URLs) | 
| BC breaks? | No
| Deprecations? | No

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

#### ToDo

- [ ] Write more functional tests to guarantee the charts and counts work as they should.
- [ ] Fix the command as it seems it does not count contacts properly. There is a functional test that shows it.
- [ ] Implement the Pending Count that has been added after this PR was submitted. It shows bad counts because of it after rebase.

[//]: # ( Required: )
#### Description:
**TL;DR Make big campaigns practical.**

![](https://i.imgur.com/jVj89GR.png)

If you have a campaign with 20 events, and 1700 leads a day, that's over a million new rows, per month and if the campaign is still active the queries on the campaign view page will take several minutes on a good day.

Even with various recent improvements, the problem is still growing exponentially as you scale up. This PR aims to fix the problem permanently by summarizing historical campaign data by hours (as services like Google Analytics do). The view will behave exactly the same (same charts, same tabular data, no delays), but will have drastically fewer rows to crunch through to generate the data (no joins, just one flat table with some sums). 

A command line function `mautic:campaign:summarize` is provided to perform an optional one-time backfill of the summary data. This might be needed by someone who wants to turn on campaign summary data *and* needs it to be retroactive. The command runs backward through time (current data is usually more pertinent), filling in missing summary data till it's done. It handles about 10.5 million campaign event log entries per hour in my testing. It's not super speedy with 200+mil. events, but it's a one-time thing. This command can also be used to true-up data in case you are experiencing exceptions within your campaign cron tasks.

I've merged in and refactored work from #6021 (Optionally display campaign data by date range) for a number of reasons. This allows you to enable "Use date range" so that all data on the Campaign view respects the date range provided. This makes the view faster and more useful for long-running campaigns. This also adds support for query caching (for whoever may use that feature by modifying their config).

**We've been running a variation of this PR (as a patch) in production via mautic-eb since Sept 2018.** It's pretty critical for us, as we have campaigns with ~28 million leads in them. Viewing such a campaign without this is literally impossible.

#### Steps to test this PR:
1. Go to https://mautibox.com/6651/s/config/edit and Campaign Settings.
2. Enable "Use summary statistics" and "Use date range"
3. Progress a contact through any campaign.
4. Go to the campaign https://mautibox.com/6651/s/campaigns and view the Campaign Statistics chart. It should display the events exactly as you would expect.
5. View the Campaign Actions/Conditions/Decisions tabs (if there are any) and they should show the same results they always have as well.
6. Pick a date range with no contacts and hit apply. You'll see the campaign Actions/Conditions/Decisions/Contacts tabs disappear or contain no content. These tabs now respect the date range provided.
